packages/node/src/behaviors/thermostat/AtomicWriteHandler.ts

```diff
@@ -0,0 +1,410 @@
+/**
+ * @license
+ * Copyright 2022-2025 Matter.js Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { ActionContext, Behavior, ClusterBehavior, type ClusterState, ValueSupervisor } from "#behavior/index.js";
+import { Thermostat } from "#clusters/thermostat";
+import { Endpoint } from "#endpoint/Endpoint.js";
+import { BasicSet, Environment, Environmental, InternalError, Logger, ObserverGroup, serialize } from "#general";
+import { ClusterModel, DataModelPath } from "#model";
+import {
+    AccessControl,
+    assertRemoteActor,
+    Fabric,
+    FabricManager,
+    hasRemoteActor,
+    PeerAddress,
+    Subject,
+    Val,
+} from "#protocol";
+import { AttributeId, NodeId, Status, StatusResponse, StatusResponseError } from "#types";
+import { AtomicWriteState } from "./AtomicWriteState.js";
+
+const logger = Logger.get("AtomicWriteHandler");
+
+/**
+ * Handles atomic write handling according to Matter definitions.
+ * The implementation tries to be generic, but is currently only used by the Thermostat cluster, so the atomic write
+ * types are imported from there.
+ *
+ * The logic requires that the cluster behavior implements the following additional events:
+ * * `${attributeName}$AtomicChanging` - emitted when an attribute is changed as part of an atomic write, before the value
+ *   is actually changed. Receives the new value, the old value and the action context as parameters.
+ * * `${attributeName}$AtomicChanged` - emitted when an attribute is changed as part of an atomic write, after the value
+ *   is actually changed. Receives the new value, the old value and the action context as parameters.
```

Comment:
Is there any reason we need additional events?  Could we use `$Changing` for notification of atomic change and `$Changed` for notification after commit like normal?

Just thinking since you can't do a non-atomic mutation of an atomic attribute the `$Changing` event is kind of useless as a separate thing.

For `$AtomicChanged` it looks like you're using that for as an additional validation step.  We might consider `validateChanges()` method on the behavior or something where you can validate the state as a whole prior to commit (atomic or otherwise).  That seems like it would be general purpose; it would handle the atomic use case but also allow you to perform inter-field validation during normal commit without having to add a listener for every interdependent attribute.